File: app/api/routes/images.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.services.image_service import ImageService
from app.schemas.image import (
    PresignUploadRequest,
    PresignUploadResponse,
    ImageCreateRequest,
    ImageResponse,
    ConfirmUploadRequest,
)
from app.api.dependencies import get_current_user
from app.models.user import User

router = APIRouter(prefix="/images", tags=["images"])


@router.post("/presign-upload", response_model=PresignUploadResponse, status_code=status.HTTP_201_CREATED)
async def presign_upload(
    request: PresignUploadRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Generate presigned S3 upload URL.
    
    Returns a presigned URL that can be used to upload the image directly to S3.
    The client should:
    1. Call this endpoint to get the presigned URL
    2. Upload the image to the presigned URL using PUT request
    3. Call POST /images to confirm the upload and store metadata
    """
    image_service = ImageService(db)
    return await image_service.generate_presigned_upload_url(
        user_id=current_user.id,
        request=request,
    )


@router.post("", response_model=ImageResponse, status_code=status.HTTP_201_CREATED)
async def create_image(
    image_data: ImageCreateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Create image record after upload.
    
    This endpoint should be called after the image has been uploaded to S3 using
    the presigned URL from /presign-upload.
    
    It validates the upload and stores metadata in the database.
    """
    image_service = ImageService(db)
    image = await image_service.create_image_record(
        user_id=current_user.id,
        image_data=image_data,
    )
    return image


@router.post("/confirm-upload", status_code=status.HTTP_200_OK)
async def confirm_upload(
    request: ConfirmUploadRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Confirm image upload and update status.
    
    Alternative endpoint to confirm upload if image was created via presign-upload.
    Body: {"s3_key": "images/user_id/uuid.jpg"}
    """
    image_service = ImageService(db)
    await image_service.confirm_upload(current_user.id, request.s3_key)
    return {"message": "Upload confirmed"}

File: web/lib/api/images.ts
import { apiClient } from './client';

export interface PresignUploadRequest {
  filename: string;
  content_type: string;
  file_size: number;
}

export interface PresignUploadResponse {
  upload_url: string;
  s3_key: string;
  image_id: number;
  expires_in: number;
}

export interface ImageCreateRequest {
  s3_key: string;
  filename: string;
  original_filename: string;
  content_type: string;
  file_size: number;
  metadata?: Record<string, any>;
}

export interface ImageResponse {
  id: number;
  user_id: number;
  filename: string;
  original_filename: string;
  content_type: string;
  file_type: 'jpg' | 'jpeg' | 'png' | 'heic';
  file_size: number;
  s3_key: string;
  s3_bucket: string;
  s3_url?: string;
  status: 'pending' | 'uploaded' | 'failed';
  metadata?: string;
  created_at: string;
  updated_at?: string;
}

export const imagesApi = {
  presignUpload: async (data: PresignUploadRequest): Promise<PresignUploadResponse> => {
    return apiClient.post<PresignUploadResponse>('/images/presign-upload', data);
  },

  createImage: async (data: ImageCreateRequest): Promise<ImageResponse> => {
    return apiClient.post<ImageResponse>('/images', data);
  },

  confirmUpload: async (s3Key: string): Promise<{ message: string }> => {
    return apiClient.post('/images/confirm-upload', { s3_key: s3Key });
  },
};




